cloud_Azure/terraform/module/examples/multiple_accounts_multiple_resource_group/profiles_populator.py

```diff
@@ -0,0 +1,505 @@
+        log.error("Failed to consent permissions to Service Principal")
+        return False
+    return True
+
+
+def save_profiles(file_path: str, profiles: List[AzureProfile]) -> bool:
+    config = configparser.ConfigParser()
+    for profile in profiles:
+        section = profile.name
+        config.add_section(section)
+        config[section]["subscription_id"] = profile.subscription_id
+        config[section]["tenant_id"] = profile.tenant_id
+        config[section]["principal_id"] = profile.principal_id
+        config[section]["principal_secret"] = profile.principal_secret
+        config[section]["location"] = profile.location
+        config[section]["resource_group_names"] = ",".join(profile.resource_group_names)
+        config[section]["storage_account_names"] = ""
+
+    try:
+        with open(file_path, "w", encoding="utf-8") as configfile:
+            config.write(configfile)
+        log.info("Saved profiles to '%s'", file_path)
+        return True
+    except OSError:
+        log.exception("Failed to save profiles to '%s'", file_path)
+        return False
+
+
+def profile_exists(profile_name: str, profiles: List[AzureProfile]) -> bool:
+    return find_profile(profile_name, profiles) is not None
+
+
+def find_profile(profile_name: str, profiles: List[AzureProfile]) -> Optional[int]:
+    for i, profile in enumerate(profiles):
+        if profile.name == profile_name:
+            return i
+    return None
+
+
+def find_secret(principal_id: str, profiles: List[AzureProfile]) -> Optional[str]:
+    for profile in profiles:
+        if profile.principal_id == principal_id and profile.principal_secret != "":
+            if likely_valid_service_principal_secret(profile.principal_secret):
+                log.info("Secret for Service Principal ID '%s' found in profile '%s'", principal_id, profile.name)
+                return profile.principal_secret
+            log.warning(
+                "Secret for Service Principal ID '%s' in profile '%s' is set, but has incorrect format",
+                principal_id,
+                profile.name,
+            )
+    log.warning("Secret for Service Principal ID '%s' not found", principal_id)
+    return None
+
+
+def likely_valid_service_principal_secret(secret: str) -> bool:
+    SERVICE_PRINCIPAL_SECRET_LENGTH = 34
+    return len(secret) == SERVICE_PRINCIPAL_SECRET_LENGTH
+
+
+def insert_or_overwrite_profile(profiles: List[AzureProfile], profile: AzureProfile) -> None:
+    index = find_profile(profile.name, profiles)
+    if index is not None:
+        profiles[index] = profile
+    else:
+        profiles.append(profile)
+
+
+def backup_file(file_path: str) -> bool:
+    """Create file backup with time stamp suffix in the name"""
+
+    if not os.path.exists(file_path):
+        log.info("Source file '%s' doesn't exist. Skipping backup", file_path)
+        return True  # nothing to backup -> OK
+
+    backup_time = datetime.today().strftime("%Y-%m-%d_%H%M%S")
```

Comment:
Safer to keep all versions, the profiles file content translates to terraform infrastructure, and it can be difficult to manage the already created infrastructure when profiles data is removed/overriden by mistake

Reply:
OK, we will though have to clearly document that:
a) we are creating a new directory for backups
b) every successful execution of the tool results in a copy of `profiles.ini` to be stored in the directory

Reply:
updated README